R/conditions.R

```diff
@@ -435,37 +439,91 @@ cnd_body.vctrs_error_recycle_incompatible_size <- function(cnd, ...) {
 
 # Names -------------------------------------------------------------------
 
-stop_names <- function(message, class, locations, ...) {
-  stop_vctrs(
-    message,
-    class = c(class, "vctrs_error_names"),
-    locations = locations,
-    ...
+new_error_names <- function(class = NULL, names) {
+  new_error_vctrs(
+    c(class, "vctrs_error_names"),
+    names = names
   )
 }
 
-stop_names_cannot_be_empty <- function(locations) {
-  stop_names(
-    "Names must not be empty.",
+stop_names_cannot_be_empty <- function(names) {
+  cnd_signal(new_error_names(
     class = "vctrs_error_names_cannot_be_empty",
-    locations = locations
-  )
+    names = names
+  ))
+}
+
+#' @export
+cnd_header.vctrs_error_names_cannot_be_empty <- function(cnd, ...) {
+  "Names can't be empty."
 }
 
-stop_names_cannot_be_dot_dot <- function(locations) {
-  stop_names(
-    "Names must not be of the form `...` or `..j`.",
+#' @export
+cnd_body.vctrs_error_names_cannot_be_empty <- function(cnd, ...) {
+  locations <- detect_empty_names(cnd$names)
+
+  if (length(locations) == 1) {
+    bullet <- glue::glue("Empty name found at location {locations}.")
+  } else {
+    bullet <- glue::glue("Empty names found at locations {ensure_full_stop(enumerate(locations))}")
+  }
+
+  bullet <- c(x = bullet)
+  format_error_bullets(bullet)
+}
+
+stop_names_cannot_be_dot_dot <- function(names) {
+  cnd_signal(new_error_names(
     class = "vctrs_error_names_cannot_be_dot_dot",
-    locations = locations
-  )
+    names = names
+  ))
+}
+
+#' @export
+cnd_header.vctrs_error_names_cannot_be_dot_dot <- function(cnd, ...) {
+  "Names can't be of the form `...` or `..j`."
 }
 
-stop_names_must_be_unique <- function(locations) {
-  stop_names(
-    "Names must be unique.",
+#' @export
+cnd_body.vctrs_error_names_cannot_be_dot_dot <- function(cnd, ...) {
+  names <- cnd$names
+
+  locations <- detect_dot_dot(names)
+  names <- names[locations]
+
+  split <- vec_group_loc(names)
+
+  info <- map2_chr(split$key, split$loc, make_names_loc_bullet)
+  bullets(info)
+}
+
+stop_names_must_be_unique <- function(names) {
+  cnd_signal(new_error_names(
     class = "vctrs_error_names_must_be_unique",
-    locations = locations
-  )
+    names = names
+  ))
+}
+
+#' @export
+cnd_header.vctrs_error_names_must_be_unique <- function(cnd, ...) {
+  "Names must be unique. Repeated names:"
```

Comment:
Should probably move `Repeated names:` in the body. It could be an `x` bullet, e.g.

```
Error: Names must be unique.
x We found these repeated names:
* ....
```

Reply:
I also wonder if "duplicate names" isn't more standard terminology than "repeated names"?

Reply:
This is what I ended up with, with indents!

```r
stop_names_must_be_unique(c("x", "x", "y", "y"))
#> Error: Names must be unique.
#> x These names are duplicated:
#>   * "x" (1 and 2)
#>   * "y" (3 and 4)
```

Reply:
hmm... I think it makes sense. What do you think about these indents @hadley and @batpigandme?

oh also the bullets are missing full stops and it might be better to make it clear that the numbers are locations:

```
#> Error: Names must be unique.
#> x These names are duplicated:
#>   * "x" at locations 1 and 2.
#>   * "y" at locations 3 and 4.
```

Reply:
I've updated to use full stops and the `"at locations"` suggestion.

I also updated `stop_names_cannot_be_dot_dot()` to this form for consistency